pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+			}
+		}
+	}
+	return &preFilterState{
+		readWriteOncePodPVCs: pvcNames,
+		conflictingPods:      conflictingPods,
+	}, nil
 }
 
-// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
-// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
-// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as UnschedulableAndUnresolvable.
-// TODO(#103132): Mark pod as Unschedulable and add preemption logic.
-func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(ctx context.Context, pod *v1.Pod) *framework.Status {
+func (pl *VolumeRestrictions) readWriteOncePodPVCsForPod(ctx context.Context, pod *v1.Pod) ([]*v1.PersistentVolumeClaim, error) {
+	pvcs := []*v1.PersistentVolumeClaim{}
 	for _, volume := range pod.Spec.Volumes {
 		if volume.PersistentVolumeClaim == nil {
 			continue
 		}
 
 		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
 		if err != nil {
-			if apierrors.IsNotFound(err) {
-				return framework.NewStatus(framework.UnschedulableAndUnresolvable, err.Error())
-			}
-			return framework.AsStatus(err)
+			return nil, err
 		}
 
 		if !v1helper.ContainsAccessMode(pvc.Spec.AccessModes, v1.ReadWriteOncePod) {
 			continue
 		}
+		pvcs = append(pvcs, pvc)
+	}
+	return pvcs, nil
+}
+
+// Checks if scheduling the pod onto this node would cause any conflicts with
+// existing volumes.
+func satisfyExistingVolumes(pod *v1.Pod, nodeInfo *framework.NodeInfo) bool {
+	for i := range pod.Spec.Volumes {
+		v := &pod.Spec.Volumes[i]
+		// fast path if there is no conflict checking targets.
+		if v.GCEPersistentDisk == nil && v.AWSElasticBlockStore == nil && v.RBD == nil && v.ISCSI == nil {
+			continue
+		}
 
-		key := framework.GetNamespacedName(pod.Namespace, volume.PersistentVolumeClaim.ClaimName)
-		if pl.sharedLister.StorageInfos().IsPVCUsedByPods(key) {
-			return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonReadWriteOncePodConflict)
+		for _, ev := range nodeInfo.Pods {
+			if isVolumeConflict(v, ev.Pod) {
+				return false
+			}
 		}
 	}
+	return true
+}
 
-	return nil
+// Checks if scheduling the pod onto this node would cause any PVC
+// ReadWriteOncePod access mode conflicts. If a pod is already scheduled on the
+// node that uses the same PVC with ReadWriteOncePod and has lower priority, it
+// can be preempted.
+func satisfyReadWriteOncePod(ctx context.Context, state *preFilterState, pod *v1.Pod) *framework.Status {
+	if state == nil {
+		return framework.NewStatus(framework.Success)
+	}
+	if pod == nil {
+		return framework.NewStatus(framework.Error)
+	}
+
+	priority := corev1helpers.PodPriority(pod)
+	code := framework.Success
+	reasons := []string{}
+
+	for _, conflictingPod := range state.conflictingPods {
+		conflictingPriority := corev1helpers.PodPriority(conflictingPod)
+		if conflictingPriority >= priority {
```

Comment:
why are you checking priority?

The preemption plugin already handles priority when simulating removal of pods.

Reply:
I didn't know, thought I might need to do this. From my understanding it sounds like (at a high-level):
- if PVC not found, return `UnschedulableAndUnresolvable`
- if there are conflicting pods, return `Unschedulable` (regardless of priority)
- else return success

Reply:
correct